Source/WebCore/platform/PublicSuffix.h

```diff
@@ -25,13 +25,33 @@
 
 #pragma once
 
+#include <wtf/HashMap.h>
+#include <wtf/HashSet.h>
+#include <wtf/text/StringHash.h>
 #include <wtf/text/WTFString.h>
 
 namespace WebCore {
 
-WEBCORE_EXPORT bool isPublicSuffix(StringView domain);
-WEBCORE_EXPORT String topPrivatelyControlledDomain(const String& domain);
-WEBCORE_EXPORT void setTopPrivatelyControlledDomain(const String& domain, const String& topPrivatelyControlledDomain);
-String decodeHostName(const String& domain);
+class PublicSuffixStore {
+public:
+    WEBCORE_EXPORT static PublicSuffixStore& singleton();
+    PublicSuffixStore() = default;
+
+    WEBCORE_EXPORT void enablePublicSuffixCache();
+    WEBCORE_EXPORT void addPublicSuffix(const String& domain);
+
+    WEBCORE_EXPORT bool isPublicSuffix(StringView domain);
+    WEBCORE_EXPORT String publicSuffix(const String& domain);
+    WEBCORE_EXPORT String topPrivatelyControlledDomain(const String& host);
+
+private:
+    bool platformIsPublicSuffix(StringView domain);
+    String platformTopPrivatelyControlledDomain(const String& host);
+
+    Lock m_HostTopPrivatelyControlledDomainCacheLock;
+    HashMap<String, String, ASCIICaseInsensitiveHash> m_hostTopPrivatelyControlledDomainCache WTF_GUARDED_BY_LOCK(m_HostTopPrivatelyControlledDomainCacheLock);
+    Lock m_publicSuffixCacheLock;
+    std::optional<HashSet<String, ASCIICaseInsensitiveHash>> m_publicSuffixCache WTF_GUARDED_BY_LOCK(m_publicSuffixCacheLock);
```

Comment:
I am not sure `cache` is the word you are going for here as this isn't something that can be invalidated and still yield correct results.

From use, it seems like this is more of an "additionalPublicSuffixes" set.